### PR TITLE
speed up accumulate, accumulateSquare, accumulateProduct and accumula…

### PR DESCRIPTION
**Pull request by vpisarev**
### What does this PR change?
-  speed up accumulate, accumulateSquare, accumulateProduct and accumulateWeighted using SIMD
- use SSE and/or AVX based on configuration
- revise the test to verify the implementation (ushort->float and float->float)
- This PR covers #6658

### Former #6798

### CODE REVIEW

**Comment by tomoaki0705** [comment]
Former #6687

File: modules/imgproc/src/accum.cpp [hunk 1]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+        int x = 0;
+
+        if (!mask)
+        {
+            len *= cn;
+            for ( ; x <= len - 4; x += 4)
+            {
+                __m128 v_src = _mm_loadu_ps(src + x);
+                __m128d v_src0 = _mm_cvtps_pd(v_src);
+                __m128d v_src1 = _mm_cvtps_pd(_mm_shuffle_ps(v_src, v_src, 0x4E));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: please, use macro `_MM_SHUFFLE` instead of magic numbers like `0x4E`

Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: it's even better to use `_mm_movehl_ps(v_src, v_src)` call.

File: modules/imgproc/src/accum.cpp [hunk 2]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: `_mm_movehl_ps`

File: modules/imgproc/src/accum.cpp [hunk 3]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+            len *= cn;
+            for ( ; x <= len - 4; x += 4)
+            {
+                __m128 v_1src  = _mm_loadu_ps(src1 + x);
+                __m128 v_2src  = _mm_loadu_ps(src2 + x);
+
+                __m128d v_1src0 = _mm_cvtps_pd(v_1src);
+                __m128d v_1src1 = _mm_cvtps_pd(_mm_shuffle_ps(v_1src, v_1src, 0xE));
+                __m128d v_2src0 = _mm_cvtps_pd(v_2src);
+                __m128d v_2src1 = _mm_cvtps_pd(_mm_shuffle_ps(v_2src, v_2src, 0xE));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: `_mm_movehl_ps` here and above.

File: modules/imgproc/src/accum.cpp [hunk 4]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+        {
+            len *= cn;
+            for ( ; x <= len - 8; x += 8)
+            {
+                __m128 v_src0 = _mm_loadu_ps(src + x);
+                __m128 v_src1 = _mm_loadu_ps(src + x + 4);
+                __m128d v_src00 = _mm_cvtps_pd(v_src0);
+                __m128d v_src01 = _mm_cvtps_pd(_mm_shuffle_ps(v_src0, v_src0, 0x0E));
+                __m128d v_src10 = _mm_cvtps_pd(v_src1);
+                __m128d v_src11 = _mm_cvtps_pd(_mm_shuffle_ps(v_src1, v_src1, 0x0E));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: `_MM_SHUFFLE`

File: modules/imgproc/src/accum.cpp [hunk 5]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+                __m128i v_int10 = _mm_unpacklo_epi16(v_int1, v_0);
+                __m128i v_int11 = _mm_unpackhi_epi16(v_int1, v_0);
+                __m128d v_src0 = _mm_cvtepi32_pd(v_int00);
+                __m128d v_src1 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int00, 0x0E));
+                __m128d v_src2 = _mm_cvtepi32_pd(v_int01);
+                __m128d v_src3 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int01, 0x0E));
+                __m128d v_src4 = _mm_cvtepi32_pd(v_int10);
+                __m128d v_src5 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int10, 0x0E));
+                __m128d v_src6 = _mm_cvtepi32_pd(v_int11);
+                __m128d v_src7 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int11, 0x0E));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: `_MM_SHUFFLE`

File: modules/imgproc/src/accum.cpp [hunk 6]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+            len *= cn;
+            for ( ; x <= len - 8; x += 8)
+            {
+                __m128i v_src  = _mm_loadu_si128((const __m128i*)(src + x));
+                __m128i v_int0 = _mm_unpacklo_epi16(v_src, v_0);
+                __m128i v_int1 = _mm_unpackhi_epi16(v_src, v_0);
+                __m128d v_src0 = _mm_cvtepi32_pd(v_int0);
+                __m128d v_src1 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int0, 0x0E));
+                __m128d v_src2 = _mm_cvtepi32_pd(v_int1);
+                __m128d v_src3 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int1, 0x0E));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: _MM_SHUFFLE

File: modules/imgproc/src/accum.cpp [hunk 7]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+    {
+        int x = 0;
+
+        if (!mask)
+        {
+            __m128i v_0 = _mm_setzero_si128();
+            len *= cn;
+            for ( ; x <= len - 16; x += 16)
+            {
+                __m128i v_src  = _mm_loadu_si128((const __m128i*)(src + x));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: I think it's better to process only 8 bytes and use `_mm_loadl_epi64`

File: modules/imgproc/src/accum.cpp [hunk 8]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+            len *= cn;
+            for ( ; x <= len - 8; x += 8)
+            {
+                __m128i v_src  = _mm_loadu_si128((const __m128i*)(src + x));
+                __m128i v_int0 = _mm_unpacklo_epi16(v_src, v_0);
+                __m128i v_int1 = _mm_unpackhi_epi16(v_src, v_0);
+                __m128d v_src0 = _mm_cvtepi32_pd(v_int0);
+                __m128d v_src1 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int0, 0x0E));
+                __m128d v_src2 = _mm_cvtepi32_pd(v_int1);
+                __m128d v_src3 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int1, 0x0E));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: `_MM_SHUFFLE`

File: modules/imgproc/src/accum.cpp [hunk 9]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+                _mm_storeu_ps(dst + x + 4, _mm_add_ps(_mm_loadu_ps(dst + x + 4), v_src1));
+            }
+        }
+        else if (cn == 1)
+        {
+            __m128i v_65535 = _mm_set1_epi16(-1);
+
+            for ( ; x <= len - 8; x += 8)
+            {
+                __m128i v_mask = _mm_loadu_si128((const __m128i*)(mask + x));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: incorrect code. You loads 16 bytes, but it's allowed only 8. Use `_mm_loadl_epi64`

Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: and you don't use 8-15 bytes anyway.

File: modules/imgproc/src/accum.cpp [hunk 10]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+        int x = 0;
+
+        if (!mask)
+        {
+            __m128i v_0 = _mm_setzero_si128();
+            len *= cn;
+            for ( ; x <= len - 16; x += 16)
+            {
+                __m128i v_1src  = _mm_loadu_si128((const __m128i*)(src1 + x));
+                __m128i v_2src  = _mm_loadu_si128((const __m128i*)(src2 + x));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: what's about 8 instead of 18 bytes? Loop iteration looks too havy

File: modules/imgproc/src/accum.cpp [hunk 11]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+            for ( ; x <= len - 8; x += 8)
+            {
+                __m128i v_1src  = _mm_loadu_si128((const __m128i*)(src1 + x));
+                __m128i v_2src  = _mm_loadu_si128((const __m128i*)(src2 + x));
+                __m128i v_1int0 = _mm_unpacklo_epi16(v_1src, v_0);
+                __m128i v_1int1 = _mm_unpackhi_epi16(v_1src, v_0);
+                __m128i v_2int0 = _mm_unpacklo_epi16(v_2src, v_0);
+                __m128i v_2int1 = _mm_unpackhi_epi16(v_2src, v_0);
+                __m128d v_src0 = _mm_mul_pd(_mm_cvtepi32_pd(v_1int0),_mm_cvtepi32_pd(v_2int0));
+                __m128d v_src1 = _mm_mul_pd(_mm_cvtepi32_pd(_mm_shuffle_epi32(v_1int0, 0x0E)),_mm_cvtepi32_pd(_mm_shuffle_epi32(v_2int0, 0x0E)));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: use `_MM_SHUFFLE` macro.

File: modules/imgproc/src/accum.cpp [hunk 12]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+        __m128d v_alpha = _mm_set1_pd(alpha);
+        __m128d v_beta = _mm_set1_pd(1.0f - alpha);
+        __m128i v_0 = _mm_setzero_si128();
+
+        if (!mask)
+        {
+            len *= cn;
+            for ( ; x <= len - 16; x += 16)
+            {
+                __m128i v_src  = _mm_loadu_si128((const __m128i*)(src + x));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: 8 bytes instead of 16?

File: modules/imgproc/src/accum.cpp [hunk 13]
```diff
@@ -83,6 +83,1434 @@ struct AccW_SIMD
...
+            len *= cn;
+            for ( ; x <= len - 8; x += 8)
+            {
+                __m128i v_src = _mm_loadu_si128((const __m128i*)(src + x));
+                __m128i v_int0 = _mm_unpacklo_epi16(v_src, v_0);
+                __m128i v_int1 = _mm_unpackhi_epi16(v_src, v_0);
+                __m128d v_src00 = _mm_cvtepi32_pd(v_int0);
+                __m128d v_src01 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int0, 0x0E));
+                __m128d v_src10 = _mm_cvtepi32_pd(v_int1);
+                __m128d v_src11 = _mm_cvtepi32_pd(_mm_shuffle_epi32(v_int1, 0x0E));
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/accum.cpp]: `_MM_SHUFFLE`

**Comment by tomoaki0705** [comment]
@ilya-lavrenov , thank you for your review and your time.
I removed the magic number using the macro, and I modified the 16bytes load to 8bytes.
Could you again have a look, please ?

**Comment by ilya-lavrenov** [comment]
thanks :+1:

**Comment by mshabunin** [comment]
@tomoaki0705 , please squash all commits into one.

**Comment by tomoaki0705** [comment]
Thank you @mshabunin and @ilya-lavrenov .
I squashed the commits.

**Pull request merged**